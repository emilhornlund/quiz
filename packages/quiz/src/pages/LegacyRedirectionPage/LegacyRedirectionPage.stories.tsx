import type { Meta, StoryObj } from '@storybook/react'
import { withRouter } from 'storybook-addon-remix-react-router'

import LegacyRedirectionPage from './LegacyRedirectionPage'

const meta = {
  component: LegacyRedirectionPage,
  decorators: [withRouter],
  parameters: { layout: 'fullscreen' },
  // Donâ€™t navigate in stories by default
  args: { disableRedirect: true },
  argTypes: {
    disableRedirect: { control: 'boolean' },
    overrideMigrationToken: { control: 'text' },
  },
} satisfies Meta<typeof LegacyRedirectionPage>

export default meta
type Story = StoryObj<typeof meta>

export const WithoutToken = {
  name: 'Without migrationToken',
  args: { overrideMigrationToken: undefined },
} satisfies Story

export const WithToken = {
  name: 'With migrationToken',
  args: {
    overrideMigrationToken: 'jU4n2n9eC-8GEZhk8NcApcfNQF9xO0yQOeJUZQk4w-E',
  },
} satisfies Story
