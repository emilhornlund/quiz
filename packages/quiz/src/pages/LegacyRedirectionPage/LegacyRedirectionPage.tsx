import { faPaperPlane } from '@fortawesome/free-solid-svg-icons'
import React, { FC, useCallback, useEffect, useMemo } from 'react'
import { useCountdown, useLocalStorage } from 'usehooks-ts'

import { Button, Page, Typography } from '../../components'
import { REDIRECT_TARGET_HOST } from '../../utils/constants.ts'

type LegacyRedirectionPageProps = {
  /** Show UI but never navigate (useful for Storybook) */
  disableRedirect?: boolean
  /** Storybook-only helper: overrides localStorage('migrationToken') on first render */
  overrideMigrationToken?: string | undefined
}

const LegacyRedirectionPage: FC<LegacyRedirectionPageProps> = ({
  overrideMigrationToken,
  disableRedirect = false,
}) => {
  const [migrationToken] = useLocalStorage<string | undefined>(
    'migrationToken',
    overrideMigrationToken,
  )

  const redirectionLink = useMemo(() => {
    const searchParams = new URLSearchParams()

    if (migrationToken) {
      searchParams.set('migrationToken', migrationToken)
    }

    const searchParamString =
      searchParams.size > 0 ? `?${searchParams.toString()}` : ''

    return `https://${REDIRECT_TARGET_HOST}${searchParamString}`
  }, [migrationToken])

  const [count, { startCountdown }] = useCountdown({
    countStart: 10,
    intervalMs: 1000,
  })

  useEffect(() => {
    startCountdown()
  }, [startCountdown])

  const handleRedirection = useCallback(() => {
    if (!disableRedirect) {
      console.log(`Redirecting to '${redirectionLink}'.`)
      window.location.replace(redirectionLink)
    }
  }, [disableRedirect, redirectionLink])

  useEffect(() => {
    if (count === 0) {
      handleRedirection()
    }
  }, [count, handleRedirection])

  return (
    <Page hideLogin>
      <Typography variant="title">We’ve moved—come with us!</Typography>

      <Typography variant="subtitle">
        Your new home is&nbsp;<strong>{REDIRECT_TARGET_HOST}</strong>
      </Typography>

      <Typography variant="text" size="medium">
        {migrationToken ? (
          <>We’ve saved your old profile and we’ll bring it along.</>
        ) : (
          <>Fresh start—no past activity found on this device.</>
        )}
      </Typography>

      <Typography variant="text" size="medium" aria-live="polite">
        Warping you over in&nbsp;<strong>{count}s</strong>…
      </Typography>

      <Button
        id="redirect-button"
        type="button"
        kind="call-to-action"
        icon={faPaperPlane}
        iconPosition="leading"
        onClick={handleRedirection}>
        Take Me There Now!
      </Button>
    </Page>
  )
}

export default LegacyRedirectionPage
