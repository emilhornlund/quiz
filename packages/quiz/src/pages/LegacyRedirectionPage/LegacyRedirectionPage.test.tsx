import { cleanup, render, screen } from '@testing-library/react'
import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'

// IMPORTANT: mock the same specifiers the component imports
vi.mock('../../utils/constants.ts', () => ({
  REDIRECT_TARGET_HOST: 'example.com',
}))

// Default mocks for usehooks-ts; individual tests can override
let startCountdownSpy = vi.fn()
let mockedCount = 10

vi.mock('usehooks-ts', async () => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const actual = await vi.importActual<any>('usehooks-ts')
  return {
    ...actual,
    useLocalStorage: vi.fn((key: string, initial?: unknown) => {
      const raw = window.localStorage.getItem(key)
      return [raw ?? (initial as unknown), vi.fn()] as const
    }),
    useCountdown: vi.fn(() => {
      return [mockedCount, { startCountdown: startCountdownSpy }] as const
    }),
  }
})

import LegacyRedirectionPage from './LegacyRedirectionPage'

// helper to render with router
const renderWithRouter = (ui: React.ReactNode) =>
  render(<MemoryRouter>{ui}</MemoryRouter>)

describe('LegacyRedirectionPage', () => {
  beforeEach(() => {
    cleanup()
    window.localStorage.clear()
    startCountdownSpy = vi.fn()
    mockedCount = 10
    vi.spyOn(console, 'log').mockImplementation(() => {})
  })

  afterEach(() => {
    vi.restoreAllMocks()
  })

  it('renders title and shows target host in subtitle', () => {
    const { container } = renderWithRouter(<LegacyRedirectionPage />)

    expect(screen.getByText(/We’ve moved—come with us!/i)).toBeInTheDocument()

    expect(screen.getByText(/Your new home is/i).textContent).toContain(
      'example.com',
    )

    expect(container).toMatchSnapshot()
  })

  it('shows the “saved profile” message when a migration token exists and builds the link with the token', () => {
    const token = 'abc123'
    window.localStorage.setItem('migrationToken', token)

    const { container } = renderWithRouter(<LegacyRedirectionPage />)

    expect(
      screen.getByText(/We’ve saved your old profile/i),
    ).toBeInTheDocument()

    const link = screen.getByRole('link', { name: /Take me there now/i })
    expect(link).toHaveAttribute(
      'href',
      `https://example.com?migrationToken=${token}`,
    )
    expect(link).toHaveAttribute('target', '_blank')
    expect(link).toHaveAttribute('rel', 'noreferrer')

    expect(container).toMatchSnapshot()
  })

  it('shows the “fresh start” message when no migration token exists and builds the link without query', () => {
    const { container } = renderWithRouter(<LegacyRedirectionPage />)

    expect(
      screen.getByText(/Fresh start—no past activity found on this device./i),
    ).toBeInTheDocument()

    const link = screen.getByRole('link', { name: /Take me there now/i })
    expect(link).toHaveAttribute('href', 'https://example.com')

    expect(container).toMatchSnapshot()
  })

  it('starts the countdown on mount', () => {
    const { container } = renderWithRouter(<LegacyRedirectionPage />)
    expect(startCountdownSpy).toHaveBeenCalledTimes(1)

    expect(container).toMatchSnapshot()
  })

  it('announces the remaining seconds in the aria-live region', () => {
    mockedCount = 7
    const { container } = renderWithRouter(<LegacyRedirectionPage />)

    const live = screen.getByText(/Warping you over in/i)
    expect(live.textContent).toContain('7s')

    expect(container).toMatchSnapshot()
  })

  it('redirects with window.location.replace and logs when the count hits 0', () => {
    const token = 'xyz'
    window.localStorage.setItem('migrationToken', token)

    // Force the hook to return 0 immediately so the effect triggers on first render
    mockedCount = 0

    // Mock window.location.replace safely in JSDOM
    const originalLocation = window.location
    const replaceMock = vi.fn()
    Object.defineProperty(window, 'location', {
      writable: true,
      value: {
        ...originalLocation,
        replace: replaceMock,
      },
    })

    const { container } = renderWithRouter(<LegacyRedirectionPage />)

    const expectedUrl = `https://example.com?migrationToken=${token}`
    expect(console.log).toHaveBeenCalledWith(`Redirecting to '${expectedUrl}'.`)
    expect(replaceMock).toHaveBeenCalledWith(expectedUrl)

    // Restore original location to avoid side-effects on other tests
    Object.defineProperty(window, 'location', { value: originalLocation })

    expect(container).toMatchSnapshot()
  })
})
