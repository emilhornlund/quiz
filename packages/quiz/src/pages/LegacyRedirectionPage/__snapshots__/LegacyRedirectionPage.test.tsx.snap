// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LegacyRedirectionPage > announces the remaining seconds in the aria-live region 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      />
    </div>
    <div
      class="content fullWidth centerAlign"
    >
      <div
        class="typography title full"
      >
        We’ve moved—come with us!
      </div>
      <div
        class="typography subtitle full"
      >
        Your new home is 
        <strong>
          example.com
        </strong>
      </div>
      <div
        class="typography text medium"
      >
        Fresh start—no past activity found on this device.
      </div>
      <div
        class="typography text medium"
      >
        Warping you over in 
        <strong>
          7
          s
        </strong>
        …
      </div>
      <div
        class="buttonInputContainer buttonInputKindCallToAction"
      >
        <button
          data-testid="test-redirect-button-button"
          id="redirect-button"
          name="redirect-button"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-paper-plane "
            data-icon="paper-plane"
            data-prefix="fas"
            role="img"
            viewBox="0 0 576 512"
          >
            <path
              d="M536.4-26.3c9.8-3.5 20.6-1 28 6.3s9.8 18.2 6.3 28l-178 496.9c-5 13.9-18.1 23.1-32.8 23.1-14.2 0-27-8.6-32.3-21.7l-64.2-158c-4.5-11-2.5-23.6 5.2-32.6l94.5-112.4c5.1-6.1 4.7-15-.9-20.6s-14.6-6-20.6-.9L229.2 276.1c-9.1 7.6-21.6 9.6-32.6 5.2L38.1 216.8c-13.1-5.3-21.7-18.1-21.7-32.3 0-14.7 9.2-27.8 23.1-32.8l496.9-178z"
              fill="currentColor"
            />
          </svg>
          <span>
            Take Me There Now!
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`LegacyRedirectionPage > redirects with window.location.replace and logs when the count hits 0 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      />
    </div>
    <div
      class="content fullWidth centerAlign"
    >
      <div
        class="typography title full"
      >
        We’ve moved—come with us!
      </div>
      <div
        class="typography subtitle full"
      >
        Your new home is 
        <strong>
          example.com
        </strong>
      </div>
      <div
        class="typography text medium"
      >
        We’ve saved your old profile and we’ll bring it along.
      </div>
      <div
        class="typography text medium"
      >
        Warping you over in 
        <strong>
          0
          s
        </strong>
        …
      </div>
      <div
        class="buttonInputContainer buttonInputKindCallToAction"
      >
        <button
          data-testid="test-redirect-button-button"
          id="redirect-button"
          name="redirect-button"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-paper-plane "
            data-icon="paper-plane"
            data-prefix="fas"
            role="img"
            viewBox="0 0 576 512"
          >
            <path
              d="M536.4-26.3c9.8-3.5 20.6-1 28 6.3s9.8 18.2 6.3 28l-178 496.9c-5 13.9-18.1 23.1-32.8 23.1-14.2 0-27-8.6-32.3-21.7l-64.2-158c-4.5-11-2.5-23.6 5.2-32.6l94.5-112.4c5.1-6.1 4.7-15-.9-20.6s-14.6-6-20.6-.9L229.2 276.1c-9.1 7.6-21.6 9.6-32.6 5.2L38.1 216.8c-13.1-5.3-21.7-18.1-21.7-32.3 0-14.7 9.2-27.8 23.1-32.8l496.9-178z"
              fill="currentColor"
            />
          </svg>
          <span>
            Take Me There Now!
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`LegacyRedirectionPage > renders title and shows target host in subtitle 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      />
    </div>
    <div
      class="content fullWidth centerAlign"
    >
      <div
        class="typography title full"
      >
        We’ve moved—come with us!
      </div>
      <div
        class="typography subtitle full"
      >
        Your new home is 
        <strong>
          example.com
        </strong>
      </div>
      <div
        class="typography text medium"
      >
        Fresh start—no past activity found on this device.
      </div>
      <div
        class="typography text medium"
      >
        Warping you over in 
        <strong>
          10
          s
        </strong>
        …
      </div>
      <div
        class="buttonInputContainer buttonInputKindCallToAction"
      >
        <button
          data-testid="test-redirect-button-button"
          id="redirect-button"
          name="redirect-button"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-paper-plane "
            data-icon="paper-plane"
            data-prefix="fas"
            role="img"
            viewBox="0 0 576 512"
          >
            <path
              d="M536.4-26.3c9.8-3.5 20.6-1 28 6.3s9.8 18.2 6.3 28l-178 496.9c-5 13.9-18.1 23.1-32.8 23.1-14.2 0-27-8.6-32.3-21.7l-64.2-158c-4.5-11-2.5-23.6 5.2-32.6l94.5-112.4c5.1-6.1 4.7-15-.9-20.6s-14.6-6-20.6-.9L229.2 276.1c-9.1 7.6-21.6 9.6-32.6 5.2L38.1 216.8c-13.1-5.3-21.7-18.1-21.7-32.3 0-14.7 9.2-27.8 23.1-32.8l496.9-178z"
              fill="currentColor"
            />
          </svg>
          <span>
            Take Me There Now!
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`LegacyRedirectionPage > shows the “fresh start” message when no migration token exists 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      />
    </div>
    <div
      class="content fullWidth centerAlign"
    >
      <div
        class="typography title full"
      >
        We’ve moved—come with us!
      </div>
      <div
        class="typography subtitle full"
      >
        Your new home is 
        <strong>
          example.com
        </strong>
      </div>
      <div
        class="typography text medium"
      >
        Fresh start—no past activity found on this device.
      </div>
      <div
        class="typography text medium"
      >
        Warping you over in 
        <strong>
          10
          s
        </strong>
        …
      </div>
      <div
        class="buttonInputContainer buttonInputKindCallToAction"
      >
        <button
          data-testid="test-redirect-button-button"
          id="redirect-button"
          name="redirect-button"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-paper-plane "
            data-icon="paper-plane"
            data-prefix="fas"
            role="img"
            viewBox="0 0 576 512"
          >
            <path
              d="M536.4-26.3c9.8-3.5 20.6-1 28 6.3s9.8 18.2 6.3 28l-178 496.9c-5 13.9-18.1 23.1-32.8 23.1-14.2 0-27-8.6-32.3-21.7l-64.2-158c-4.5-11-2.5-23.6 5.2-32.6l94.5-112.4c5.1-6.1 4.7-15-.9-20.6s-14.6-6-20.6-.9L229.2 276.1c-9.1 7.6-21.6 9.6-32.6 5.2L38.1 216.8c-13.1-5.3-21.7-18.1-21.7-32.3 0-14.7 9.2-27.8 23.1-32.8l496.9-178z"
              fill="currentColor"
            />
          </svg>
          <span>
            Take Me There Now!
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`LegacyRedirectionPage > shows the “saved profile” message when a migration token exists 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      />
    </div>
    <div
      class="content fullWidth centerAlign"
    >
      <div
        class="typography title full"
      >
        We’ve moved—come with us!
      </div>
      <div
        class="typography subtitle full"
      >
        Your new home is 
        <strong>
          example.com
        </strong>
      </div>
      <div
        class="typography text medium"
      >
        We’ve saved your old profile and we’ll bring it along.
      </div>
      <div
        class="typography text medium"
      >
        Warping you over in 
        <strong>
          10
          s
        </strong>
        …
      </div>
      <div
        class="buttonInputContainer buttonInputKindCallToAction"
      >
        <button
          data-testid="test-redirect-button-button"
          id="redirect-button"
          name="redirect-button"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-paper-plane "
            data-icon="paper-plane"
            data-prefix="fas"
            role="img"
            viewBox="0 0 576 512"
          >
            <path
              d="M536.4-26.3c9.8-3.5 20.6-1 28 6.3s9.8 18.2 6.3 28l-178 496.9c-5 13.9-18.1 23.1-32.8 23.1-14.2 0-27-8.6-32.3-21.7l-64.2-158c-4.5-11-2.5-23.6 5.2-32.6l94.5-112.4c5.1-6.1 4.7-15-.9-20.6s-14.6-6-20.6-.9L229.2 276.1c-9.1 7.6-21.6 9.6-32.6 5.2L38.1 216.8c-13.1-5.3-21.7-18.1-21.7-32.3 0-14.7 9.2-27.8 23.1-32.8l496.9-178z"
              fill="currentColor"
            />
          </svg>
          <span>
            Take Me There Now!
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`LegacyRedirectionPage > starts the countdown on mount 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="header"
    >
      <button
        class="logo"
      >
        <img
          alt="Klurigo"
          class="icon"
          src="/src/assets/images/klurigo-icon.svg"
        />
        <span
          class="text"
        >
          Klurigo
        </span>
      </button>
      <div
        class="side"
      />
    </div>
    <div
      class="content fullWidth centerAlign"
    >
      <div
        class="typography title full"
      >
        We’ve moved—come with us!
      </div>
      <div
        class="typography subtitle full"
      >
        Your new home is 
        <strong>
          example.com
        </strong>
      </div>
      <div
        class="typography text medium"
      >
        Fresh start—no past activity found on this device.
      </div>
      <div
        class="typography text medium"
      >
        Warping you over in 
        <strong>
          10
          s
        </strong>
        …
      </div>
      <div
        class="buttonInputContainer buttonInputKindCallToAction"
      >
        <button
          data-testid="test-redirect-button-button"
          id="redirect-button"
          name="redirect-button"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-paper-plane "
            data-icon="paper-plane"
            data-prefix="fas"
            role="img"
            viewBox="0 0 576 512"
          >
            <path
              d="M536.4-26.3c9.8-3.5 20.6-1 28 6.3s9.8 18.2 6.3 28l-178 496.9c-5 13.9-18.1 23.1-32.8 23.1-14.2 0-27-8.6-32.3-21.7l-64.2-158c-4.5-11-2.5-23.6 5.2-32.6l94.5-112.4c5.1-6.1 4.7-15-.9-20.6s-14.6-6-20.6-.9L229.2 276.1c-9.1 7.6-21.6 9.6-32.6 5.2L38.1 216.8c-13.1-5.3-21.7-18.1-21.7-32.3 0-14.7 9.2-27.8 23.1-32.8l496.9-178z"
              fill="currentColor"
            />
          </svg>
          <span>
            Take Me There Now!
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
